js/ui/grid_core/filtering/utils.js

```diff
@@ -0,0 +1,152 @@
+// @ts-check
+
+import { Deferred } from '../../../core/utils/deferred';
+import DataSource from '../../../data/data_source';
+import { normalizeDataSourceOptions } from '../../../data/data_source/utils';
+import variableWrapper from '../../../core/utils/variable_wrapper';
+import { isFunction, isString } from '../../../core/utils/type';
+import gridCoreUtils from '../ui.grid_core.utils';
+
+function normalizeGroupingLoadOptions(group) {
+    if(!Array.isArray(group)) {
+        group = [group];
+    }
+
+    return group.map((item, i) => {
+        if(isString(item)) {
+            return {
+                selector: item,
+                isExpanded: i < group.length - 1,
+            };
+        }
+
+        return item;
+    });
+}
+
+/**
+ *
+ * @param {import('../columns_controller/types').Lookup} lookup
+ */
+export function normalizeLookupDataSource(lookup) {
+    let lookupDataSourceOptions;
+    if(lookup.items) {
+        lookupDataSourceOptions = lookup.items;
+    } else {
+        lookupDataSourceOptions = lookup.dataSource;
+        if(isFunction(lookupDataSourceOptions) && !variableWrapper.isWrapped(lookupDataSourceOptions)) {
+            lookupDataSourceOptions = lookupDataSourceOptions({});
+        }
+    }
+
+    return normalizeDataSourceOptions(lookupDataSourceOptions);
+}
```

Comment:
Let's reduce the complexity of this function:
```suggestion
export function normalizeLookupDataSource(lookup) {
    if(lookup.items) {
        return normalizeDataSourceOptions(lookup.items);
    } else if(
        isFunction(lookup.dataSource) 
        && !variableWrapper.isWrapped(lookup.dataSource)
    ) {
        return normalizeDataSourceOptions(lookup.dataSource({}));
    }

    return normalizeDataSourceOptions(lookup.dataSource);
}
```